vpr/src/pack/prepack.h

```diff
@@ -6,16 +6,175 @@
 
 #ifndef PREPACK_H
 #define PREPACK_H
+
+#include <algorithm>
+#include <map>
 #include <unordered_map>
-#include "atom_netlist_fwd.h"
-#include "arch_types.h"
 #include "vpr_types.h"
+#include "vtr_range.h"
+
+class AtomNetlist;
+class AtomBlockId;
+struct t_molecule_stats;
+struct t_logical_block_type;
+
+/**
+ * @brief Class that performs prepacking.
+ *
+ * This class maintains the prepacking state, allowing the use of molecules
+ * (prepacked atoms) while this object exists.
```

Comment:
Say molecules currently come from pack patterns in the architecture file, and give an example of a molecule, e.g. a 3-bit carry chain in most architectures would turn into a molecule containing the 3 atoms forming that carry chain.

Reply:
Also comment on whether every atom is part of a molecule or not -- I think they all are, where atoms that are not part of larger pre-pack structures like carry chains still form a single-atom molecule.

Reply:
Sounds good. I will update the doc. I think you are correct that every atom block will be part of a molecule, that is at least what I am seeing in AP and what the code is suggesting.

There is one confusing aspect I could not document: Based on the API, it is possible for a block to be within multiple molecules (hence `get_atom_molecules` returns a range). I have yet to find out when this happens. In AP I just put an assert to check that the number of molecules per block is 1, and on simple architectures this assert is never hit (I have not tried complicated architectures/circuits yet). Do you happen to know when this case occurs? I can always update this part of the documentation (or force it to always be 1) later.

Reply:
The code was very hard to understand, but I think the blocks being part of multiple molecules has something to do with carry chains.

Reply:
It could be used for things like a carry chain plus LUT/FF.  Say you have a 3-bit carry chain where every bit also drives a FF and the pack pattern say pack LUT/add->FF connections.  Now you could in theory have 4 molecules (partially overlapping) where each carry bit is part of two of them.

LUT_wcarry -> LUT_wcarry -> LUT_wcarry
  |                         |                       |
 FF                      FF                     FF

But I think early code merges those pack patterns into one larger molecule so in the end this is likely a single 6-atom molecule anyway, except perhaps during molecule creation?

Reply:
AH! That makes a lot of sense! Especially if this is only used internally. The code that makes use of the range was super odd to me. This makes sense now. What I will do is try to add an assert and see if the late-stage of prepacking is cleaning it up; and if so I will make the accessor return a single molecule instead. It will just make the API easier to work with (and make more sense).

Reply:
If not, I will document this in the API.